### PR TITLE
Fix bucket reports in BlobStoreStats to support TTL_UPDATE and UNDELETE

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
This PR
1. Fixes bugs that fail test cases
2. Add two delta buckets for log segment to deal with delete and expiration separately
3. Add support for TTL_UPDATE and UNDELETE
4. Add more test cases to test DELETE, TTL_UPDATE and UNDELETE.

### CODE REVIEW

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:577]: Why TtlUpdate is also counted here?

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:616]: minor: can we call this `indexValuePriorToNewDelete` or the more descriptive name?

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:577]: Because we use findKey to get the value, and in findKey, we are not returning TTL_UPDATE record.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:833]: Where is `latestValue` used? In this `if` block, no matter what the `value` is, the StoreException will be thrown and `latestValue` is never used.

File: ambry-store/src/main/java/com/github/ambry/store/ScanResults.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/ScanResults.java:82]: Let's call this `containerDeltaBuckets`.  Btw, what does `bucketCount` comprise? Is bucketCount = 1 baseBucket + number of delta buckets?

File: ambry-store/src/main/java/com/github/ambry/store/ScanResults.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/ScanResults.java:135]: `containerForecastEndTimeMs`,  `logSegmentForecastEndTimeMsForExpired` and  `logSegmentForecastEndTimeMsForDeleted` all seem to be extending another `bucketSpanInMs` from last bucket time. Can you explain why? We can talk this offline if needed.

File: ambry-store/src/main/java/com/github/ambry/store/ScanResults.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/ScanResults.java:101]: I think the maps in this class are better to concurrent map as IndexScanner and QueueProcessor may concurrently update the map.

**Review by jsjtzyy — COMMENTED** [review]
Continuing to review.

File: ambry-store/src/main/java/com/github/ambry/store/ScanResults.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/ScanResults.java:101]: QueueProcessor shouldn't be running at the same time with IndexScanner. IndexScanner would change `isScanning` to true while scanning the index files. And when the `isScanning` is true, QueueProcessor won't do anything.

File: ambry-store/src/main/java/com/github/ambry/store/ScanResults.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/ScanResults.java:82]: sure.

So when bucketCount is 1, we only have baseBucket.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:833]: you are right, will update.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:539]: will update

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:145]: I am thinking about 24 buckets. One bucket for 10 minutes and 4 hours. I am willing to extend to 8 or 10 hours. So 24, 48 or 60 are all good options for me.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:577]: because then the TTL_UPDATE flag is true, value.isPut will be false. findKey has this weird logic where it will populate TTL_UPDATE flag even if we are not trying to find TTL_UPDATE record. 

So when we have PUT, TTL_UPDATE and DELETE for this blob and we call findKey(blobId, [INDEX_TYPE.PUT]),  we will get a PUT value with TTL_UPDATE flag being true. And if we do value.isPut, it will return false.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:575]: will update

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:274]: I am actually confused about this `scanLock`. If IndexScanner is running (which means `isScanning == true`), current thread is able to acquire and hold `scanLock` (because it can be released by IndexScanner at line 1089). After that, current thread can be blocked by `waitCondition` (without releasing lock), however, IndexScanner has to re-acquire the  lock before signaling the `waitCondition` (see line 1141 - 1146).  It seems like a deadlock except for `waitCondition` can time out eventually. Is this intended?

**Review by jsjtzyy — COMMENTED** [review]
a few more comments

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:274]: waitCondition is a condition on scanLock, when calling waitCondition.await, it will release the scan lock. This is pretty much how the condition and lock work.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:1239]: `removeFinalStateOnPut` will remove the final state of a storeKey when encountering PUT record, since PUT record is the first record for the storeKey. This is for reducing memory usage. We don't want to keep final states of all store key in the map throughout the iteration of all index segment. 

However, in this method, we will process same index segment twice, first time for calculating valid size of log segment, second time for calculating valid size for container. So when calculating for log segment, we don't remove the final state  since later we would need the final state when calculating for containers.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:659]: updated.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:709]: that was a mistake. updated.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:811]: if we are under this if statement, then we know the current IndexValue is a DELETE and the previous IndexValue is also a DELETE, the IndexValue history can go like this, PUT, DELETE(v0), DELETE(v1).

When we see the DELETE(v0), we will 1) add size of DELETE to base bucket, then 2)subtract size of PUT int the DELETE's operation time bucket. 
When we see the DELETE(v1), we know it will invalidate DELETE(v0). So one way to do this is to revert what DELETE(v0) did before, so we 1)subtract size of DELETE from base bucket, then 3)add size of PUT in the DELETE(v0)'s operation time bucket. After doing this, we can treat DELETE(v0) just like it never happened. Then we move on to deal with DELETE(v1).

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:275]: Got it, I didn't notice `waitCondition` will release the lock.

**Comment by zzmao** [comment]
BlobStore changes looks good to me.

After spending some time to read the blobstats code, but I feel I am not qualified to review blobstats code 😢 . @SophieGuo410  probably a good reviewer for blobstats change.

**Comment by justinlin-linkedin** [comment]
@SophieGuo410 adding you as third reviewer if you are interested.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:577]: I see. Thanks for explanation. I think it probably justifies a comment here.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:812]: Correct, this is what I thought about. My suggestion is to make the comments more accurate:
`Added it's size to the first bucket in the scan result` -> `Added it's size to the base bucket`;
`Subtract previous DELETE's size from bucket in the scan result` -> `Subtract previous DELETE's size from base bucket`;
Also add one sentence in the end something like:
```
After recovering changes made by previous DELETE, we update both base bucket and operation time bucket for new DELETE.
```

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:886]: I understand this part and just want to confirm again about UNDELETE. The UNDELETE operation always checks if original PUT has expired. So, if we reach here, it means original PUT hasn't expired and should be subtracted, correct?

I am asking this because I am thinking about a case where original PUT expires before DELETE and there is an UNDELETE in the end. In this case, original PUT is subtracted twice but I think this case would never happen.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java]: I reword the comments, please take another look, see if it's good for you.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:899]: Let's consider a case: PUT, DELETE, UNDELETE and PUT expires after UNDELETE op time. Based on my understanding, it will go through following steps:
```
 Operation  | Base bucket | Delete Op bucket | Undelete Op bucket | Expired Bucket
    DELETE   | + D size        |  - P size                 |                                     |  + P size
 UNDELETE | - D + U size  |  + P size                |                                     |  + P size
```
Looks like the expired bucket adds PUT record size twice, did I misunderstand anything?

**Review by jsjtzyy — COMMENTED** [review]
Two more comments, other code looks good.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:899]: if the current IndexValue is UNDELETE and the previous IndexValue is DELETE, then we will first revert the change made by DELETE, so for UNDELETE, at Expired Bucket, it will do -P size, not, +P size.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:886]: Your first statement is correct, and there is no chance when the PUT is expired before DELETE and UNDELETE would succeeds. If we have PUT and DELETE for the blob, but PUT is already expired, then UNDELETE will fail since it can't undelete an expired PUT.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:988]: I understand the piece of code is to check whether TtlUpdate is valid or not. If PUT and TtlUpdate are in same log segment and final state is delete, then TtlUpdate is not valid. 
I wonder if we can search the log segment of TtlUpdate only instead of findKey from the very beginning of whole PersistentIndex.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:983]: Since the `indexValue` is not from findKey() method, then I believe this can be simplified to `indexValue.isTtlUpdate()`.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:1032]: I wonder if it's worthwhile to add an additional check in this "for" statement. That is, checking if there is IndexScanner is ongoing. 
```
for (int i = 0; i < entryCount && !cancelled &&  !isScanning; i++)
```
It's possible `QueueProcessor` starts before `IndexScanner`. When `IndexScanner` may begin while `QueueProcessor` is still ongoing.

**Review by jsjtzyy — APPROVED** [review]
LGTM after remaining comments are addressed.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:988]: that's a good idea, updated.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:983]: A DELETE or UNDELETE IndexValue can still return true when calling isTtlUpdate(), so there we have to do this kinda check.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:1032]: that's right, will put a check just in case.

**Review by zzmao — APPROVED** [review]
LGTM

**Pull request merged**
